### PR TITLE
HDDS-3153. Create REST API to serve Recon Dashboard and integrate with UI in Recon.

### PR DESCRIPTION
**Pull request by vivekratnavel**
## What changes were proposed in this pull request?

- Add REST api endpoint to serve cluster state (/api/v1/clusterState)
- Integrate UI with clusterState API
- Remove dangling javadoc comment in LICENSE block from all Recon files
- Add unit and integration tests

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3153

## How was this patch tested?

unit tests, integration tests and manual tests using docker-compose.

Screen Shot 2020-03-14 at 2 55 22 PM

### CODE REVIEW

**Comment by vivekratnavel** [comment]
@avijayanhwx @elek Please review

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/DatanodesCount.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlElement;
+
+/**
+ * Metadata object that contains datanode counts based on its state.
+ */
+@XmlAccessorType(XmlAccessType.FIELD)
+public class DatanodesCount {
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/DatanodesCount.java]: IMO this class is redundant. We can just capture the num datanodes (healthy and total) as 2 longs in the response.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/PipelineMetadata.java [hunk 2]
```diff
@@ -167,9 +149,19 @@ public PipelineMetadata build() {
       Preconditions.checkNotNull(datanodes);
       Preconditions.checkNotNull(replicationType);
 
-      return new PipelineMetadata(pipelineId, status, leaderNode, datanodes,
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/PipelineMetadata.java]: Why was this changed from using the constructor? Setting the class field directly is non standard.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    AtomicInteger healthyDatanodes = new AtomicInteger();
+    int containers = this.containerManager.getContainerIDs().size();
+    int pipelines = this.pipelineManager.getPipelines().size();
+    long volumes;
+    long buckets;
+    long keys;
+    AtomicLong capacity = new AtomicLong(0L);
+    AtomicLong used = new AtomicLong(0L);
+    AtomicLong remaining = new AtomicLong(0L);
+    datanodeDetails.forEach(datanode -> {
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: We can use SCMNodeManager#getNodeCount(NodeState)  to get the number of nodes in a specific state.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        healthyDatanodes.get());
+    ClusterStateResponse.Builder builder = ClusterStateResponse.newBuilder();
+    try {
+      volumes = omMetadataManager.getVolumeTable().getEstimatedKeyCount();
+      builder.setVolumes(volumes);
+    } catch (Exception ex) {
+      LOG.error("Unable to get Volumes count in ClusterStateResponse.", ex);
+    }
+    try {
+      buckets = omMetadataManager.getBucketTable().getEstimatedKeyCount();
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: Nit. We can directly use builder.set(omm.getTable(). getEstimatedKeyCount()). We don't need local variables. (Same for keys, volumes, pipelines, containers etc)

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  public Response getClusterState() {
+    List<DatanodeDetails> datanodeDetails = nodeManager.getAllNodes();
+    AtomicInteger healthyDatanodes = new AtomicInteger();
+    int containers = this.containerManager.getContainerIDs().size();
+    int pipelines = this.pipelineManager.getPipelines().size();
+    long volumes;
+    long buckets;
+    long keys;
+    AtomicLong capacity = new AtomicLong(0L);
+    AtomicLong used = new AtomicLong(0L);
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: Why AtomicLong? Simple long may be enough.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    AtomicLong remaining = new AtomicLong(0L);
+    datanodeDetails.forEach(datanode -> {
+      NodeState nodeState = nodeManager.getNodeState(datanode);
+      SCMNodeStat nodeStat = nodeManager.getNodeStat(datanode).get();
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: Can we use SCMNodeManager#getStats here? It is supposed to give the aggregate stats from all nodes.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+   * @return {@link Response}
+   */
+  @GET
+  public Response getClusterState() {
+    List<DatanodeDetails> datanodeDetails = nodeManager.getAllNodes();
+    int containers = this.containerManager.getContainerIDs().size();
+    int pipelines = this.pipelineManager.getPipelines().size();
+    int healthyDatanodes = nodeManager.getNodeCount(NodeState.HEALTHY);
+    SCMNodeStat stats = nodeManager.getStats();
+    DatanodeStorageReport storageReport =
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: Unused variable.

File: hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestEndpoints.java [hunk 8]
```diff
@@ -305,4 +419,39 @@ public void testGetPipelines() throws Exception {
...
+    // then container is not added until the next container report is processed
+    SCMHeartbeatRequestProto heartbeatRequestProto =
+        SCMHeartbeatRequestProto.newBuilder()
+            .setContainerReport(containerReportsProto)
+            .setDatanodeDetails(datanodeDetailsProto)
+            .build();
+    reconScm.getDatanodeProtocolServer()
+        .sendHeartbeat(heartbeatRequestProto);
+
+    LambdaTestUtils.await(30000, 2000, () -> {
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestEndpoints.java]: Line 440 - 454 has some some repetitive elements in all 3 tests. Only the condition is different. Maybe we can use a helper function to do this and remove duplicate code.

`  private void waitAndCheckConditionAfterHeartbeat(Callable<Boolean> check) throws Exception {
    // if container report is processed first, and pipeline does not exist
    // then container is not added until the next container report is processed
    SCMHeartbeatRequestProto heartbeatRequestProto =
        SCMHeartbeatRequestProto.newBuilder()
            .setContainerReport(containerReportsProto)
            .setDatanodeDetails(datanodeDetailsProto)
            .build();
    reconScm.getDatanodeProtocolServer()
        .sendHeartbeat(heartbeatRequestProto);
    LambdaTestUtils.await(30000, 2000, check);
  } `

It can be invoked by
`    waitAndCheckConditionAfterHeartbeat(() -> {
      Response response1 = clusterStateEndpoint.getClusterState();
      ClusterStateResponse clusterStateResponse1 =
          (ClusterStateResponse) response1.getEntity();
      return (clusterStateResponse1.getContainers() == 1);
    });`

I am OK to do this in the next JIRA if needed.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java (hunk 7 above)
Comment (vivekratnavel) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: `storageReport` is used in setStorageReport of ClusterStateResponse.Builder in L102.

File: hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestEndpoints.java (hunk 8 above)
Comment (vivekratnavel) [reply hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestEndpoints.java]: Agreed. I can take care of this refactor in another JIRA.

Comment (avijayanhwx) [reply hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestEndpoints.java]: Thanks!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java (hunk 7 above)
Comment (avijayanhwx) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ClusterStateEndpoint.java]: Sorry, I referenced the wrong storageReport. There is an unused one in TestEndpoints#testGetClusterState.

**Comment by avijayanhwx** [comment]
LGTM +1 pending CI

File: hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestEndpoints.java (hunk 8 above)
Comment (vivekratnavel) [reply hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestEndpoints.java]: Refactored as per the suggestion in the latest commit.

**Comment by vivekratnavel** [comment]
The failing test is not related to this patch and fails due to timeout issue.

**Comment by avijayanhwx** [comment]
Thank you for this patch @vivekratnavel.

**Comment by vivekratnavel** [comment]
Thanks @avijayanhwx for the review and commit.
